In [11]:
# ADITIONAL EXERCISE 1
from collections import defaultdict
from collections import Counter

SEED_NUM = 4                       # last digit of your student number
STUDENT_MAJOR = "BSECE"            # current degree program
FAVORITE_ARTIST = "ERASERHEADS"     # Favorite local artist or band
LOCAL_HAZARD = "TYPHOON"        # A local emergency event, e.g. TYPHOON, FLOOD, EARTHQUAKE

CONTROL_NUM = max(3, SEED_NUM)

stream = [
    ("TUP Manila", "BSECE"),
    ("TUP Taguig", f"BSME_v{CONTROL_NUM}"),
    ("TUP Manila", f"BSME_v{CONTROL_NUM}"),
    ("TUP Manila", "BSECE"),
    ("TUP Visayas", STUDENT_MAJOR),
    ("TUP Taguig", "BSECE"),
    ("TUP Manila", "BSECE")
]

stream.extend([
    ("TUP Manila", "BSCE"),
    ("TUP Manila", "BSCE"),
    ("TUP Manila", "BSCS"),
    ("TUP Taguig", "BSECE"),
    ("TUP Manila", "BSCS"),
    ("TUP Visayas", "BSCE"),
    ("TUP Manila", f"BSIT_v{CONTROL_NUM}")
])

campus_data = defaultdict(list)

for campus, program in stream:
    campus_data[campus].append(program)

campus_counts = {}

for campus in campus_data:
    campus_counts[campus] = Counter(campus_data[campus])

top_program = campus_counts["TUP Manila"].most_common(1)

print(campus_counts)
print(top_program)

{'TUP Manila': Counter({'BSECE': 3, 'BSCE': 2, 'BSCS': 2, 'BSME_v4': 1, 'BSIT_v4': 1}), 'TUP Taguig': Counter({'BSECE': 2, 'BSME_v4': 1}), 'TUP Visayas': Counter({'BSECE': 1, 'BSCE': 1})}
[('BSECE', 3)]


In [12]:
# ADITIONAL EXERCISE 2
CONTROL_NUM = 4
FAVORITE_ARTIST = "Shirebound"

festival = {
    "Ben&Ben", "SB19", "Bini", "Eraserheads",
    FAVORITE_ARTIST, "Zild",
    f"Indie Artist {CONTROL_NUM}"
}

user_a = {"Ben&Ben", "Bini", "Maki", "Dionela", FAVORITE_ARTIST}

user_b = {"SB19", "Eraserheads", "Zild",
          f"Indie Artist {CONTROL_NUM}", "Parokya ni Edgar"}

new_artists = {"Arthur Nery", "TJ Monterde", "Cup of Joe, Over October"}  # must match CONTROL_NUM
festival.update(new_artists)

def jaccard_similarity(set1, set2):
    intersection = set1 & set2
    union = set1 | set2
    return len(intersection), len(union), (len(intersection) / len(union)) * 100

int_a, union_a, jaccard_a = jaccard_similarity(user_a, festival)

int_b, union_b, jaccard_b = jaccard_similarity(user_b, festival)

dealbreakers_a = user_a - festival

print("Total Artists in Expanded Festival:", len(festival))

print("\nUser A:")
print("Intersection:", int_a)
print("Union:", union_a)
print("Jaccard Similarity (%):", jaccard_a)

print("\nUser B:")
print("Jaccard Similarity (%):", jaccard_b)

print("\nUser A Dealbreakers:", dealbreakers_a)

Total Artists in Expanded Festival: 10

User A:
Intersection: 3
Union: 12
Jaccard Similarity (%): 25.0

User B:
Jaccard Similarity (%): 36.36363636363637

User A Dealbreakers: {'Maki', 'Dionela'}


In [14]:
# ADITIONAL EXERCISE 3
from collections import deque, Counter

CONTROL_NUM = 4
LOCAL_HAZARD = "FIRE"

burst = [
    (CONTROL_NUM, "WEATHER"),
    (CONTROL_NUM + 1, "TRAFFIC"),
    (CONTROL_NUM + 2, "WEATHER"),
    (CONTROL_NUM + 3, "WEATHER"),
    (CONTROL_NUM + 4, LOCAL_HAZARD),
    (CONTROL_NUM + 5, "WEATHER"),
    (CONTROL_NUM + 6, "TRAFFIC")
]

extra_alerts = []
start_time = CONTROL_NUM + 7

for i in range(CONTROL_NUM + 2):
    # you can vary categories for realism
    category = ["WEATHER", "TRAFFIC", LOCAL_HAZARD][i % 3]
    extra_alerts.append((start_time + i, category))

burst.extend(extra_alerts)

dq = deque(maxlen=5)

weather_check = None
local_hazard_check = None

for alert in burst:
    timestamp, category = alert
    
    categories = [c for _, c in dq]
    count = Counter(categories)
    
    if count[category] >= 2:
        appended = False
    else:
        dq.append(alert)
        appended = True

    if timestamp == CONTROL_NUM + 3 and category == "WEATHER":
        weather_check = "Y" if appended else "N"
    
    if category == LOCAL_HAZARD:
        if local_hazard_check is None:
            local_hazard_check = "Y" if appended else "N"

final_sum = sum(ts for ts, _ in dq)

print("Total Burst Size:", len(burst))
print(f'Did "WEATHER" at Timestamp {CONTROL_NUM + 3} append?:', weather_check)
print("Did LOCAL_HAZARD append?:", local_hazard_check)
print("Final Deque Length:", len(dq))
print("Final Sum of Timestamps in Deque:", final_sum)
print("Exact Final Deque Output:", list(dq))

Total Burst Size: 13
Did "WEATHER" at Timestamp 7 append?: N
Did LOCAL_HAZARD append?: Y
Final Deque Length: 5
Final Sum of Timestamps in Deque: 60
Exact Final Deque Output: [(8, 'FIRE'), (10, 'TRAFFIC'), (13, 'FIRE'), (14, 'WEATHER'), (15, 'TRAFFIC')]


ADITIONAL EXERCISE #1
{'TUP Manila': Counter({'BSECE': 3, 'BSCE': 2, 'BSCS': 2, 'BSME_v4': 1, 'BSIT_v4': 1}), 'TUP Taguig': Counter({'BSECE': 2, 'BSME_v4': 1}), 'TUP Visayas': Counter({'BSECE': 1, 'BSCE': 1})}
[('BSECE', 3)]

ADITIONAL EXERCISE #2
Total Artists in Expanded Festival: 10

User A:
Intersection: 3
Union: 12
Jaccard Similarity (%): 25.0

User B:
Jaccard Similarity (%): 36.36363636363637

User A Dealbreakers: {'Maki', 'Dionela'}

ADITIONAL EXERCISE #3
Total Burst Size: 13
Did "WEATHER" at Timestamp 7 append?: N
Did LOCAL_HAZARD append?: Y
Final Deque Length: 5
Final Sum of Timestamps in Deque: 60
Exact Final Deque Output: [(8, 'FIRE'), (10, 'TRAFFIC'), (13, 'FIRE'), (14, 'WEATHER'), (15, 'TRAFFIC')]